src/Microsoft.AspNetCore.Razor.Language/Legacy/RazorSyntaxTreeExtensions.cs

```diff
@@ -0,0 +1,130 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.Language.Legacy
+{
+    internal static class RazorSyntaxTreeExtensions
+    {
+        public static IReadOnlyList<ClassifiedSpanInternal> GetClassifiedSpans(this RazorSyntaxTree syntaxTree)
+        {
+            if (syntaxTree == null)
+            {
+                throw new ArgumentNullException(nameof(syntaxTree));
+            }
+
+            var spans = Flatten(syntaxTree);
+
+            var result = new ClassifiedSpanInternal[spans.Count];
+            for (var i = 0; i < spans.Count; i++)
+            {
+                var span = spans[i];
+                result[i] = new ClassifiedSpanInternal(
+                    new SourceSpan(
+                        span.Start.FilePath ?? syntaxTree.Source.FilePath,
+                        span.Start.AbsoluteIndex,
+                        span.Start.LineIndex,
+                        span.Start.CharacterIndex,
+                        span.Length),
+                    new SourceSpan(
+                        span.Parent.Start.FilePath ?? syntaxTree.Source.FilePath,
+                        span.Parent.Start.AbsoluteIndex,
+                        span.Parent.Start.LineIndex,
+                        span.Parent.Start.CharacterIndex,
+                        span.Parent.Length),
+                    span.Kind,
+                    span.Parent.Type,
+                    span.EditHandler.AcceptedCharacters);
+            }
+
+            return result;
+        }
+
+        public static IReadOnlyList<TagHelperSpanInternal> GetTagHelperSpans(this RazorSyntaxTree syntaxTree)
+        {
+            if (syntaxTree == null)
+            {
+                throw new ArgumentNullException(nameof(syntaxTree));
+            }
+
+            var results = new List<TagHelperSpanInternal>();
+
+            var toProcess = new List<Block>();
+            var blockChildren = new List<Block>();
+            toProcess.Add(syntaxTree.Root);
+
+            for (var i = 0; i < toProcess.Count; i++)
+            {
+                var blockNode = toProcess[i];
+                if (blockNode is TagHelperBlock tagHelperNode)
+                {
+                    results.Add(new TagHelperSpanInternal(
+                        new SourceSpan(
+                            tagHelperNode.Start.FilePath ?? syntaxTree.Source.FilePath,
+                            tagHelperNode.Start.AbsoluteIndex,
+                            tagHelperNode.Start.LineIndex,
+                            tagHelperNode.Start.CharacterIndex,
+                            tagHelperNode.Length),
+                        tagHelperNode.Binding));
+                }
+
+                // collect all child blocks and inject into toProcess as a single InsertRange
+                foreach (var curNode in blockNode.Children)
```

Comment:
I know this is probably just a move of this code, but can we update the silly names?